File: leetcode_75/queue/dota2_senate_649.java
import java.util.*;

class Solution {
    public String predictPartyVictory(String senate) {
        int n = senate.length();
        var qD = new ArrayDeque<Integer>();
        var qR = new ArrayDeque<Integer>();

        for (int i = 0; i < n; i++) {
            if (senate.charAt(i) == 'D') {
                qD.addLast(i);
            } else {
                qR.addLast(i);
            }
        }

        while (!qD.isEmpty() && !qR.isEmpty()) {
            var d = qD.pollFirst();
            var r = qR.pollFirst();

            if (d < r) {
                qD.addLast(d + n);
            } else {
                qR.addLast(r + n);
            }
        }

        return qD.isEmpty() ? "Radiant" : "Dire";
    }
}

File: leetcode_75/queue/dota2_senate_649.py
from collections import deque

class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        n = len(senate)
        qD, qR = deque(), deque()

        for i, ch in enumerate(senate):
            (qD if ch == 'D' else qR).append(i)
        # 這邊先建立分別理者的 index 對列，因為沒有被ban的選手，是可以繼續在場上的，直到下一次

        while qD and qR:
            d, r = qD.popleft(), qR.popleft()
            if d < r:
                qD.append(d + n)   # D 先出手，ban 一名 R
            else:
                qR.append(r + n)   # R 先出手，ban 一名 D
                
        return "Dire" if qD else "Radiant"
